### DISCUSSION TITLE
Basset not work for multi-tenancy structure

### DISCUSSION THREAD

**Discussion by ben-cheng565** [Laravel-Backpack/community-forum#1234] [Q&A (Help)] [answered]
In a multi-tenancy project, I recently upgraded the backpack to v6, but I have no idea how to configure APP_URL in env file. Since each tenant has a different URL, so packages cannot be loaded using basset.

[a comment by an automated account is not shown]

**Reply by siberfx** [reply]
you can predefine in AppServiceProvider according to domain like;

`$tenant = TenantModel::current();`
`Config::set('app.url', $tenant->domain)`

or if you want to modify the .env file, you can use package to modify the dotenv and add it to your worker when a new tenant has been created, so after succesful tenant creating you can modify env file, cheers!

**Reply by ben-cheng565** [reply]
Thanks, @siberfx. I have checked your answer, but it's not available to get current tenant data in AppServiceProvider. I also tried to configure the app.url somewhere else, but it looks like Backpack does not identify the new value.

**Reply by siberfx** [reply]
could you please share some pieces of your tenancy system about? so I can try to help you with that

**Reply by ben-cheng565** [reply]
Thanks. My project uses stancl/tenancy for multi-tenancy management, and the APP_URL in .env file has a value as the base URL, then each tenant has a dynamic URL generated from APP_URL, (https://subdomain.APP_URL). 
After upgrading Backpack to 6, which utilizes Basset to load resources, but Basset requires a fixed APP_URL, each backpack page cannot get resource files, as shown in the screenshot below.
<img width="383" alt="Screenshot 2024-11-19 at 2 39 21 PM" src="https://github.com/user-attachments/assets/2ed73011-4c3b-4ccc-a12f-1576b5849613">

I have got a solution from https://github.com/Laravel-Backpack/basset/issues/139#issuecomment-2374368512https://github.com/Laravel-Backpack/basset/issues/139#issuecomment-2374368512

**Reply by karandatwani92** [reply]
Hey @ben-cheng565 How are you?

Can you share the solution that worked for you? 
Some code and steps would help me write the tutorial on this and help others. Thanks 🙏

**Reply by ben-cheng565** [reply, marked as answer]
Hey @karandatwani92 , I am glad to share it to help others with a similar case.

Step 1, Run the command to generate the basset config file if it does not exist
`php artisan vendor:publish --provider="Backpack\Basset\BassetServiceProvider"`
Then change 'disk' => env('BASSET_DISK', 'public') to 'disk' => env('BASSET_DISK', 'basset')

Step 2, Add basset to the disks array in the filesystems.php
```
        'basset' => [
            'driver' => 'local',
            'root' => storage_path('app/public'),
            'url' => '/storage',      # set url without prefix for the case that central app has a different URL
            'visibility' => 'public',
            'throw' => false,
        ],
```

Step 3, Modify AppServiceProvider file to reset the disk link for tenants:
```
public function boot()
    {
        Event::listen(TenancyBootstrapped::class, function (TenancyBootstrapped $event) {
            # Your code here
            
            # Reset the basset disk link to the tenant's storage
            config()->set('backpack.basset.disk', 'basset');
            config()->set('filesystems.disks.basset.url', asset('storage'));
            app()->scoped('basset', fn() => new \Backpack\Basset\BassetManager());
        });
```

**Comment by jcastroa87** [comment]
Thanks @ben-cheng565 for sharing this information.

I will close the issue, but please feel free to re-open or create a new one if needed.

Cheers.
